### ISSUE TITLE
Convenience functionality for adding custom linear operations

### ISSUE THREAD

**Issue by mreineck** [google/jax#19179]
The `ducc0` software package (https://gitlab.mpcdf.mpg.de/mtr/ducc) implements several linear operators, like FFT and friends (which actually has been included into Jax already), but also spherical harmonic transforms, nonuniform FFTs etc.
I was looking for a convenient way of making these operations known to Jax without jumping through the hoops of https://jax.readthedocs.io/en/latest/Custom_Operation_for_GPUs.html every time; in the end I came up with something that works pretty well, although it is fairly hackish under the hood.

Adding a new linear operator to Jax (for this example's purpose, the complex-to-complex FFT, although it already exists in Jax) would look roughly like this:

```Python
import jax_linop

def c2c_operator(axes, nthreads):
    def c2cfunc(inp, out, adjoint, state):
        ducc0.fft.c2c(inp, out=out, axes=state["axes"],
                      nthreads=state["nthreads"], forward=not adjoint)
    def c2cfunc_abstract(shape, dtype, adjoint, state):
        return shape, dtype

    return jax_linop.make_linop(c2cfunc, c2cfunc_abstract, axes=tuple(axes),
                      nthreads=int(nthreads))

# create a c2c operator that transforms the first axis of its input array on 4 threads
c2c_op = c2c_operator(axes=(0,), nthreads=4)
# get the Jax function for this operator in forward direction
op_fwd = jax.jit(c2c_op)
# get the Jax function for this operator in adjoint direction
op_adj = jax.jit(c2c_op.adjoint)
```

The current state of impementation (plus a few tests) can be seen at

- https://gitlab.mpcdf.mpg.de/mtr/ducc/-/blob/more_linop/python/test/jax_linop.py?ref_type=heads
- https://gitlab.mpcdf.mpg.de/mtr/ducc/-/blob/more_linop/python/jax_pymod.cc?ref_type=heads
- https://gitlab.mpcdf.mpg.de/mtr/ducc/-/blob/more_linop/python/test/test_jax.py?ref_type=heads

The internal dependencies on `ducc` are very small; it would be easy to move the code to a standalone package.

Would there be interest in including this kind of functionality into Jax itself? If so, licensing terms shouldn't be a problem; I can switch from GPL to a more relaxed license for the files in question.

**Comment by mreineck** [comment]
I managed to separate the code completely from `ducc0`; it is now available at https://github.com/mreineck/jax_linop. Please note that this is still an early prototype for collecting feedback!

**Comment by mreineck** [comment]
Sorry ... clicked "thumbs up" on my own issue by accident and don't know how to undo it :P

**Comment by mreineck** [comment]
Here is an updated demo that works without exotic packages and shows gradient tests:

```Python
import scipy.fft
import jax_linop
import numpy as np

def fft_operator(axes):
    def fftfunc(inp, out, adjoint, state):
        # we don't want normalization, just adjointness, hence the strange norms
        if adjoint:
            out[()] = scipy.fft.ifftn(inp, axes=state["axes"], norm="forward")
        else:
            out[()] = scipy.fft.fftn(inp, axes=state["axes"], norm="backward")

    def fftfunc_abstract(shape, dtype, adjoint, state):
        return shape, dtype

    return jax_linop.make_linop(fftfunc, fftfunc_abstract, axes=tuple(axes))

# create an FFT operator that transforms the first axis of its input array,
# as well as its adjoint operator
fft_op, fft_op_adjoint = fft_operator(axes=(0,))

from jax import config
config.update("jax_enable_x64", True)

from jax.test_util import check_grads

rng = np.random.default_rng(42)
a = rng.random((100,20,30))-0.5 + 1j*(rng.random((100,20,30))-0.5)

check_grads(fft_op, (a,), order=2, modes=["fwd", "rev"], eps=1.)
check_grads(fft_op_adjoint, (a,), order=2, modes=["fwd", "rev"], eps=1.)
```

**Comment by Edenhofer** [comment]
I think this would be a great addition to JAX and we could plug Hartley transforms into JAX this way as discussed in #7490 :blush:

**Comment by quattro** [comment]
Might be worth considering the [`lineax`](https://github.com/google/lineax) library as another place for this functionality?

**Comment by Edenhofer** [comment]
@quattro IMO `lineax` is not the right place for this kind of code. The contribution proposed here has little to do with linear solves but is all about a convenient way to define custom primitives for JAX/XLA.

**Comment by Edenhofer** [comment]
@hawkinsp Would you be open to having this (or version of it) in JAX?  I'm pinging you here as you were open to the idea of bringing more of [ducc](https://gitlab.mpcdf.mpg.de/mtr/ducc) to JAX in #7490.
